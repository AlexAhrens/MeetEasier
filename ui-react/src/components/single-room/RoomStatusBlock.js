import React, { Component } from 'react';
import PropTypes from 'prop-types';

let srConfig = require('../../config/singleRoom.config.js');

class RoomStatusBlock extends Component {
  constructor(props) {
    super(props);
    this.state = {
      appointmentExists: false,
      timesPresent: false,
      nextUp: ''
    }
  }

  roomAppointmentValidation = () => {
    const { room } = this.props;

    // check if there are times in the room.Start & room.End
    // then: if the meeting is not going on now, append "Next Up: "
    if (typeof room.Appointments !== 'undefined' && room.Appointments.length > 0) {
      this.setState({
        appointmentExists: true
      });

      if (room.Appointments[0].Start && room.Appointments[0].End) {
        this.setState({
          timesPresent: true
        });

        if (room.Busy === 'true') {
          this.setState({
            nextUp: ''
          });
        }
        else {
          this.setState({
            nextUp: srConfig.nextUp + ': '
          });
        }
      }
      else {
        this.setState({
          timesPresent: false
        });
      }
    }
    else {
      this.setState({
        appointmentExists: false
      });
    }
  }

  componentDidMount = () => {
    this.roomAppointmentValidation();
  }

  render() {
    const { room } = this.state;
    const { appointmentExists } = this.state;
    return (
      <div className={room.Busy ? 'columns small-8 left-col busy' : 'columns small-8 left-col open'}>
        <div id="single-room__room-name">{room.Name}</div>
        <div id="single-room__room-status">{room.Busy ? srConfig.statusBusy : srConfig.statusAvailable}</div>
        { appointmentExists ?
          <div id="single-room__meeting-title">
            <span id="single-room__next-up">
              {nextUp}
            </span>
            {room.Appointments[0].Subject}
          </div>
        : ''
        }
        <div id="single-room__meeting-time">
          {timesPresent ?
            new Date(parseInt(room.Appointments[0].Start, 10)).toLocaleTimeString([], {weekday: 'short', hour: '2-digit', minute: '2-digit'}) + ' - ' + new Date(parseInt(item.Appointments[0].End, 10)).toLocaleTimeString([], {hour: '2-digit', minute: '2-digit'})
            : ''
          }
        </div>
        { appointmentExists ?
          <div id="single-room__meeting-organizer">
            {room.Appointments[0].Organizer}
          </div>
          : ''
        }
      </div>
    );
  }

}

RoomStatusBlock.propTypes = {
  room: Prop
}

export default RoomStatusBlock;
