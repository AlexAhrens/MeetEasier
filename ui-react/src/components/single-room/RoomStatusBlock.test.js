import RoomStatusBlock from './RoomStatusBlock';

describe('COMPONENT', () => {
  let room, details, config;
  beforeEach(() => {
    room = {
      'Roomlist':'Test Roomlist',
      'Name':'Test Room',
      'RoomAlias':'test-room',
      'Email':'email@email.com',
      'Appointments':[
         {
            'Subject':'Meeting Subject',
            'Organizer':'John Doe',
            'Start':1532966400000,
            'End':1533344400000
         }
      ],
      'Busy':true
    };

    details = {
      appointmentExists: true,
      nextUp: ''
    };

    config = {
      statusBusy: 'Busy',
      statusAvailable: 'Available'
    };
  });

  // *********************************
  // change to shallow when finished 
  // and remove this comment
  // *********************************
  it('renders correctly', () => {
    const wrapper = mount(<RoomStatusBlock room={room} details={details} config={config} />);

    expect(wrapper).toMatchSnapshot();
  });

  // Details tests

  it('renders Details', () => {
    const wrapper = shallow(<RoomStatusBlock room={room} details={details} config={config} />);

    const component = wrapper.find('Details').exists();
    expect(component).toBeTruthy();
  });

  it('renders Details with text if appointment exists', () => {
    const wrapper = mount(<RoomStatusBlock room={room} details={details} config={config} />);

    const div = wrapper.find('#single-room__details').text();
    expect(div).not.toBe('');
  });

  it('renders Details blank if appointment does not exist', () => {
    details.appointmentExists = false;
    const wrapper = mount(<RoomStatusBlock room={room} details={details} config={config} />);

    const div= wrapper.find('#single-room__details').text();
    expect(div).toBe('');
  })

  // /Details tests
  // Time tests

  it('renders Time', () => {
    const wrapper = shallow(<RoomStatusBlock room={room} details={details} config={config} />);

    const component = wrapper.find('Time').exists();
    expect(component).toBeTruthy();
  });

  it('renders Time with text if appointment exists', () => {
    const wrapper = mount(<RoomStatusBlock room={room} details={details} config={config} />);

    const div = wrapper.find('#single-room__meeting-time').text();
    expect(div).not.toBe('');
  });

  it('renders Time blank if appointment does not exist', () => {
    details.appointmentExists = false;
    const wrapper = mount(<RoomStatusBlock room={room} details={details} config={config} />);

    const div= wrapper.find('#single-room__meeting-time').text();
    expect(div).toBe('');
  })

  // /Time tests
  // Organizer tests

  it('renders Organizer', () => {
    const wrapper = shallow(<RoomStatusBlock room={room} details={details} config={config} />);

    const component = wrapper.find('Organizer').exists();
    expect(component).toBeTruthy();
  });

  it('renders Organizer with text if appointment exists', () => {
    const wrapper = mount(<RoomStatusBlock room={room} details={details} config={config} />);

    const div = wrapper.find('#single-room__meeting-organizer').text();
    expect(div).not.toBe('');
  });

  it('renders Organizer blank if appointment does not exist', () => {
    details.appointmentExists = false;
    const wrapper = mount(<RoomStatusBlock room={room} details={details} config={config} />);

    const div= wrapper.find('#single-room__meeting-organizer').text();
    expect(div).toBe('');
  })

  // /Organizer tests
});
