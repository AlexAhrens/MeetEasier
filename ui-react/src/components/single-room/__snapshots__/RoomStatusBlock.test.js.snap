// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT renders correctly 1`] = `
<div
  className="columns small-8 left-col busy"
>
  <div
    id="single-room__room-name"
  >
    Test Room
  </div>
  <div
    id="single-room__room-status"
  >
    Busy
  </div>
  <div
    id="single-room__details"
  >
    <div
      id="single-room__meeting-title"
    >
      <span
        id="single-room__next-up"
      />
      <span
        id="single-room__meeting-subject"
      >
        Meeting Subject
      </span>
    </div>
  </div>
  <div
    id="single-room__meeting-time"
  >
    Mon 9:00 AM - 6:00 PM
  </div>
  <div
    id="single-room__meeting-organizer"
  >
    John Doe
  </div>
</div>
`;
